src/utils/index_split.hpp

```diff
@@ -0,0 +1,118 @@
+//========================================================================================
+// (C) (or copyright) 2023. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef UTILS_INDEX_SPLIT_HPP_
+#define UTILS_INDEX_SPLIT_HPP_
+
+#include "basic_types.hpp"
+#include "defs.hpp"
+#include "globals.hpp"
+#include "mesh/domain.hpp"
+
+namespace parthenon {
+
+// forward declarations
+template <typename T>
+class MeshData;
+
+class IndexSplit {
+ public:
+  static constexpr int all_outer = -100;
+  static constexpr int no_outer = -200;
+  IndexSplit(MeshData<Real> *md, const IndexRange &kb, const IndexRange &jb,
+             const IndexRange &ib, const int nkp, const int njp);
+  IndexSplit(MeshData<Real> *md, IndexDomain domain, const int nkp, const int njp);
+
+  int outer_size() const { return nkp_ * njp_; }
+  KOKKOS_INLINE_FUNCTION
+  IndexRange GetBoundsK(const int p) const {
+    const auto kf = p / njp_;
+    return {kbs_ + static_cast<int>(kf * target_k_),
+            kbs_ + static_cast<int>((kf + 1) * target_k_) - 1};
```

Comment:
To be extra clear, I think this line specifically needs to 
```suggestion
            min(kbs_ + static_cast<int>((kf + 1) * target_k_) - 1,kbe_entire_)};
```